### PR TITLE
Dom: Add types

### PR DESCRIPTION
**Pull request by sirreal**
## Description

Add types to `@wordpress/dom` package.

Split out from #27344 

Part of #18838

## How has this been tested?
```
npm run build:package-types
```

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/dom/src/dom.js [hunk 1]
```diff
@@ -70,19 +79,29 @@ function isSelectionForward( selection ) {
 	return true;
 }
 
+/* eslint-disable jsdoc/valid-types */
+/**
+ * @param {Element} element Element to test
+ * @return {element is HTMLTextAreaElement | HTMLInputElement} True if element it Input or TextArea
```
Comment (ockham) [inline packages/dom/src/dom.js:85]:
```suggestion
 * @return {element is HTMLTextAreaElement | HTMLInputElement} True if element is Input or TextArea
```

**Comment by sarayourfriend** [comment]
Rebasing this PR. I'll follow up with fixes.

**Comment by sarayourfriend** [comment]
I think this PR is probably too big. I'd like to propose the following strategy instead of doing this all in one go:

1. Split `dom.js` into several smaller modules in a new `dom` folder.
2. Progressively type each module in individual PRs (fixing "spidering" issues as they come up but trying to contain changes to smaller PRs).

As it stands I think there are too many logical changes that it's not really super safe or fun to review this PR. What do y'all think @sirreal  @ockham?

**Comment by sirreal** [comment]
I agree this is quite big and an incremental approach would be better. I tried to tackle this quickly while working on compose types, but it turned out much more involved. If you have a good plan to move ahead with it, please feel free to do whatever makes sense. I don't know when I'll be able to get back to this. Even if we don't split up and opt-in files to typechecking incrementally, changes can be landed incrementally in independent PRs with the goal of enabling typechecking.

As I worked on this, there were places where it made sense to add protection. However, there are many unhandled nullable values that don't have an obvious improvement. In fact, hitting nullable values in many cases indicates the library is not being used correctly. In this case, runtime errors make a lot of sense, the library doesn't have a graceful way of dealing with some of these cases.

I'm not sure how to best handle that, one option is to liberally add type assertions and leave the runtime alone. There's a lot of appeal to that 🙂 

Another more radical option would be to add some assertions especially around the problematic nullable values and fail fast. Something like [`assertIsDefined` here](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions):

```ts
function assertIsDefined<T>(val: T): asserts val is NonNullable<T> {
  if (val === undefined || val === null) {
    throw new AssertionError(
      `Expected 'val' to be defined, but received ${val}`
    );
  }
}
```

**Comment by youknowriad** [comment]
Doing some cleanup to old PRs. This draft PR seems staled. should we close it?

**Pull request closed without merging**